{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from selenium import webdriver\n",
    "from selenium.webdriver.chrome.options import Options\n",
    "from selenium.webdriver.common.by import By\n",
    "options = Options()\n",
    "options.add_argument('--no-sandbox')\n",
    "options.add_argument('--headless')\n",
    "options.add_argument('--disable-dev-shm-usage')\n",
    "options.add_argument(\"--remote-debugging-port=9222\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('genome.txt', sep='\\t', na_values='---', dtype={'rsid':'str', 'chromosome':'object', 'position':'int', 'genotype':'str'}, comment='#', names=[\"rsid\", \"chromosome\", \"position\", \"genotype\"])\n",
    "\n",
    "data.head(50)\n",
    "data.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('genome.txt', sep='\\t', na_values='---', dtype={'rsid':'str', 'chromosome':'object', 'position':'int', 'genotype':'str'}, comment='#', names=[\"rsid\", \"chromosome\", \"position\", \"genotype\"])\n",
    "\n",
    "data['chromosome'].unique()\n",
    "data['chromosome'] = data['chromosome'].apply(lambda x:\n",
    "    re.sub(r'X', r'23', x))\n",
    "data['chromosome'] = data['chromosome'].apply(lambda x:\n",
    "    re.sub(r'MT', r'24', x))\n",
    "\n",
    "    # Figure out what to do with your Y\n",
    "data = data[data['chromosome'] != 'Y']\n",
    "data['chromosome'] = data['chromosome'].apply(lambda x: \n",
    "\tint(x))\n",
    "\n",
    "    # showing\n",
    "# data['isDigit'] = data['chromosome'].str.isdigit()\n",
    "# data = data.loc[(data['isDigit'] == False) & (data['chromosome'] == 'Y')  ]\n",
    "# data['isDigit'] = data['chromosome'].str.isdigit()\n",
    "# data = data[data['isDigit'] == False]\n",
    "# data\n",
    "\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome_dict = {1:'1', 2:'2', 3:'3', 4:'4', 5:'5', \n",
    "\t\t\t\t   6:'6', 7:'7', 8:'8', 9:'9', 10:'10', \n",
    "\t\t\t\t   11:'11', 12:'12', 13:'13', 14:'14', \n",
    "\t\t\t\t   15:'15', 16:'16', 17:'17', 18:'18', \n",
    "\t\t\t\t   19:'19', 20:'20', 21:'21', 22:'22', \n",
    "\t\t\t\t   23:'X', 24:'MT'}\n",
    "\n",
    "print(chromosome_dict)\n",
    "display(data.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove whitespace from column name\n",
    "data.rename({' rsid': 'rsid'}, axis='columns', inplace=True)\n",
    "\n",
    "#how many snps are there per chromosome?\n",
    "\n",
    "#solve manually with a for loop\n",
    "x = []\n",
    "y = []\n",
    "for k in chromosome_dict:\n",
    "    x.append(k)\n",
    "    y.append(len(data[data.chromosome ==k]))\n",
    "rsid_per_chromosome = dict(zip(x,y))\n",
    "rsid_per_chromosome\n",
    "\n",
    "#or solve with pandas\n",
    "rsid_per_chromosome_series = data.groupby('chromosome')['rsid'].count()\n",
    "rsid_per_chromosome_series.columns = ['chromosome','count']\n",
    "rsid_per_chromosome_series.plot.barh(figsize=(16,9), fontsize=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_files = glob.glob(\"./csv from snpedia/*.csv\")\n",
    "\n",
    "snp_df = pd.concat((pd.read_csv(f) for f in all_files))\n",
    "snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_df['genotype'] = snp_df['Unnamed: 0'].apply(lambda x: \n",
    "\tre.sub(r'.*([AGCT]);([AGCT])\\)', r'\\1\\2', x))\n",
    "new_cols = ['rsid', 'magnitude', 'repute', \n",
    "'summary', 'genotype']\n",
    "snp_df.columns = new_cols\n",
    "\n",
    "snp_df['rsid'] = snp_df['rsid'].map(lambda x : x.lower())\n",
    "snp_df['rsid'] = snp_df['rsid'].map(lambda x : \n",
    "\tre.sub(r'([a-z]{1,}[\\d]+)\\([agct];[agct]\\)', \n",
    "\tr'\\1', x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "null_repute = snp_df[snp_df['repute'].isnull()]\n",
    "null_summaries = snp_df[snp_df['summary'].isnull()]\n",
    "null_repute_and_summaries = pd.concat([null_repute,null_summaries]).drop_duplicates().reset_index(drop=True)\n",
    "display(null_repute_and_summaries)\n",
    "snp_df['repute'].fillna(value='Neutral', inplace=True)\n",
    "snp_df['summary'].fillna(value='None', inplace=True)\n",
    "snp_df.isna().any()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>magnitude</th>\n",
       "      <th>repute</th>\n",
       "      <th>summary</th>\n",
       "      <th>genotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4680</td>\n",
       "      <td>2.5</td>\n",
       "      <td>Good</td>\n",
       "      <td>(worrier) advantage in memory and attention tasks</td>\n",
       "      <td>AA</td>\n",
       "      <td>22</td>\n",
       "      <td>19951271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>rs601338</td>\n",
       "      <td>2.5</td>\n",
       "      <td>Good</td>\n",
       "      <td>resistance to Norovirus infection</td>\n",
       "      <td>AA</td>\n",
       "      <td>19</td>\n",
       "      <td>49206674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs4680</td>\n",
       "      <td>2.5</td>\n",
       "      <td>Good</td>\n",
       "      <td>(worrier) advantage in memory and attention tasks</td>\n",
       "      <td>AA</td>\n",
       "      <td>22</td>\n",
       "      <td>19951271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>rs9264942</td>\n",
       "      <td>2.2</td>\n",
       "      <td>Good</td>\n",
       "      <td>90% reduction in HIV viral load</td>\n",
       "      <td>CC</td>\n",
       "      <td>6</td>\n",
       "      <td>31274380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>rs1799990</td>\n",
       "      <td>2.1</td>\n",
       "      <td>Good</td>\n",
       "      <td>Resistance to vCJD (PrP 129 Met/Val heterozygo...</td>\n",
       "      <td>AG</td>\n",
       "      <td>20</td>\n",
       "      <td>4680251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>rs1805005</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Good</td>\n",
       "      <td>common in clinvar</td>\n",
       "      <td>GG</td>\n",
       "      <td>16</td>\n",
       "      <td>89985844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>rs1805007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Good</td>\n",
       "      <td>normal risk</td>\n",
       "      <td>CC</td>\n",
       "      <td>16</td>\n",
       "      <td>89986117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>rs2228479</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Good</td>\n",
       "      <td>common in clinvar</td>\n",
       "      <td>GG</td>\n",
       "      <td>16</td>\n",
       "      <td>89985940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>rs5907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Good</td>\n",
       "      <td>common in clinvar</td>\n",
       "      <td>GG</td>\n",
       "      <td>22</td>\n",
       "      <td>21134223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>i3002432</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Good</td>\n",
       "      <td>None</td>\n",
       "      <td>GG</td>\n",
       "      <td>11</td>\n",
       "      <td>46761055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>446 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsid  magnitude repute  \\\n",
       "4       rs4680        2.5   Good   \n",
       "211   rs601338        2.5   Good   \n",
       "5       rs4680        2.5   Good   \n",
       "188  rs9264942        2.2   Good   \n",
       "561  rs1799990        2.1   Good   \n",
       "..         ...        ...    ...   \n",
       "176  rs1805005        0.0   Good   \n",
       "175  rs1805007        0.0   Good   \n",
       "174  rs2228479        0.0   Good   \n",
       "173     rs5907        0.0   Good   \n",
       "793   i3002432        0.0   Good   \n",
       "\n",
       "                                               summary genotype  chromosome  \\\n",
       "4    (worrier) advantage in memory and attention tasks       AA          22   \n",
       "211                  resistance to Norovirus infection       AA          19   \n",
       "5    (worrier) advantage in memory and attention tasks       AA          22   \n",
       "188                    90% reduction in HIV viral load       CC           6   \n",
       "561  Resistance to vCJD (PrP 129 Met/Val heterozygo...       AG          20   \n",
       "..                                                 ...      ...         ...   \n",
       "176                                  common in clinvar       GG          16   \n",
       "175                                        normal risk       CC          16   \n",
       "174                                  common in clinvar       GG          16   \n",
       "173                                  common in clinvar       GG          22   \n",
       "793                                               None       GG          11   \n",
       "\n",
       "     position  \n",
       "4    19951271  \n",
       "211  49206674  \n",
       "5    19951271  \n",
       "188  31274380  \n",
       "561   4680251  \n",
       "..        ...  \n",
       "176  89985844  \n",
       "175  89986117  \n",
       "174  89985940  \n",
       "173  21134223  \n",
       "793  46761055  \n",
       "\n",
       "[446 rows x 7 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df = snp_df.merge(data, how='inner', on=['rsid', 'genotype'], suffixes=('_SNPedia', '_myDNA'))\n",
    "bad = new_df[new_df['repute'] == 'Bad'].sort_values('magnitude', ascending = False)\n",
    "neutral = new_df[new_df['repute'] == 'Neutral'].sort_values('magnitude', ascending = False)\n",
    "good = new_df[new_df['repute'] == 'Good'].sort_values('magnitude', ascending = False)\n",
    "interesting = new_df[new_df['magnitude'] > 4].sort_values('magnitude', ascending = False)\n",
    "good "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           rsid  magnitude repute  \\\n",
      "784  rs62514958        5.9    Bad   \n",
      "171  rs28930069        5.0    Bad   \n",
      "104  rs28936383        5.0    Bad   \n",
      "105  rs28936383        5.0    Bad   \n",
      "203   rs1800462        3.5    Bad   \n",
      "..          ...        ...    ...   \n",
      "37    rs1333040        0.0    Bad   \n",
      "282   rs2308327        0.0    Bad   \n",
      "489   rs8177374        0.0    Bad   \n",
      "321  rs10761659        0.0    Bad   \n",
      "272    rs440446        0.0    Bad   \n",
      "\n",
      "                                               summary genotype  chromosome  \\\n",
      "784  Non-phenylketonuria hyperphenylalaninemia geno...       GG          12   \n",
      "171                Hypokalemic periodic paralysis risk       GG           1   \n",
      "104  Limb-girdle muscular dystrophy-dystroglycanopathy       GG           4   \n",
      "105  Limb-girdle muscular dystrophy-dystroglycanopathy       GG           4   \n",
      "203             incapable of detoxifying certain drugs       CC           6   \n",
      "..                                                 ...      ...         ...   \n",
      "37   1.24x increased myocardial infarction risk, ~1...       CT           9   \n",
      "282                                             normal       AA          10   \n",
      "489  very common, but it seems better to have the m...       CC          11   \n",
      "321                       1.5x risk of Crohn's disease       AA          10   \n",
      "272       increased risk in men for biliary conditions       GG          19   \n",
      "\n",
      "      position  \n",
      "784  103240677  \n",
      "171  201022667  \n",
      "104   52895065  \n",
      "105   52895065  \n",
      "203   18143955  \n",
      "..         ...  \n",
      "37    22083404  \n",
      "282  131565170  \n",
      "489  126162843  \n",
      "321   64445564  \n",
      "272   45409167  \n",
      "\n",
      "[143 rows x 7 columns]\n",
      "           rsid  magnitude   repute  \\\n",
      "684  rs59328451        3.0  Neutral   \n",
      "683  rs57424749        3.0  Neutral   \n",
      "2     rs1426654        2.7  Neutral   \n",
      "3     rs1426654        2.7  Neutral   \n",
      "549  rs12913832        2.5  Neutral   \n",
      "..          ...        ...      ...   \n",
      "341  rs28934574        0.0  Neutral   \n",
      "342  rs28934573        0.0  Neutral   \n",
      "343  rs28933990        0.0  Neutral   \n",
      "344  rs28934876        0.0  Neutral   \n",
      "787   rs1408799        0.0  Neutral   \n",
      "\n",
      "                                               summary genotype  chromosome  \\\n",
      "684  homozygote for pachyonychia congenita Type I m...       TT          17   \n",
      "683  homozygote for pachyonychia congenita Type I m...       CC          17   \n",
      "2            probably light-skinned, European ancestry       AA          15   \n",
      "3            probably light-skinned, European ancestry       AA          15   \n",
      "549                    blue eye color, 99% of the time       GG          15   \n",
      "..                                                 ...      ...         ...   \n",
      "341                                               None       GG          17   \n",
      "342                                               None       GG          17   \n",
      "343                                               None       GG          15   \n",
      "344                                               None       TT          11   \n",
      "787                                               None       CC           9   \n",
      "\n",
      "      position  \n",
      "684   39766801  \n",
      "683   39768561  \n",
      "2     48426484  \n",
      "3     48426484  \n",
      "549   28365618  \n",
      "..         ...  \n",
      "341    7577094  \n",
      "342    7577559  \n",
      "343   89754025  \n",
      "344  118971106  \n",
      "787   12672097  \n",
      "\n",
      "[205 rows x 7 columns]\n",
      "           rsid  magnitude repute  \\\n",
      "784  rs62514958        5.9    Bad   \n",
      "104  rs28936383        5.0    Bad   \n",
      "105  rs28936383        5.0    Bad   \n",
      "171  rs28930069        5.0    Bad   \n",
      "\n",
      "                                               summary genotype  chromosome  \\\n",
      "784  Non-phenylketonuria hyperphenylalaninemia geno...       GG          12   \n",
      "104  Limb-girdle muscular dystrophy-dystroglycanopathy       GG           4   \n",
      "105  Limb-girdle muscular dystrophy-dystroglycanopathy       GG           4   \n",
      "171                Hypokalemic periodic paralysis risk       GG           1   \n",
      "\n",
      "      position  \n",
      "784  103240677  \n",
      "104   52895065  \n",
      "105   52895065  \n",
      "171  201022667  \n",
      "{'table': ['https://www.snpedia.com/index.php/rs62514958', 'https://www.snpedia.com/index.php/rs28936383', 'https://www.snpedia.com/index.php/rs28936383', 'https://www.snpedia.com/index.php/rs28930069']}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1401/4239684641.py:15: DeprecationWarning: use options instead of chrome_options\n",
      "  browser = webdriver.Chrome(chrome_options=options)\n",
      "/tmp/ipykernel_1401/4239684641.py:31: DeprecationWarning: find_element_by_* commands are deprecated. Please use find_element() instead\n",
      "  rsid = browser.find_element_by_css_selector('.firstHeading').text\n",
      "/tmp/ipykernel_1401/4239684641.py:32: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead\n",
      "  links_elements = browser.find_elements_by_partial_link_text('PMID')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "0\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Get the base URL from SNPedia\n",
    "import time\n",
    "base_url = 'https://www.snpedia.com/index.php/'\n",
    "\n",
    "genes = [bad, neutral, interesting]\n",
    "urlDictionary = {}\n",
    "for table in genes:\n",
    "    gene_urls = [base_url + rsid for rsid in table['rsid']]\n",
    "    print(gene_urls)\n",
    "    title = f'{table=}'.split('=')[0]\n",
    "    urlDictionary[title] = gene_urls\n",
    "\n",
    "print(urlDictionary)\n",
    "# Initialize Selenium\n",
    "browser = webdriver.Chrome(chrome_options=options)\n",
    "# Write a function to visit the SNPedia URLs, click through to PubMed,\n",
    "# and retrieve the info on the articles for each gene\n",
    "\n",
    "\n",
    "def scrape_abstracts(urls):\n",
    "\n",
    "    rsid_list = []\n",
    "    all_article_title = []\n",
    "    all_article_citation = []\n",
    "    all_article_authors = []\n",
    "    # all_article_abstract = []\n",
    "    all_article_links = []\n",
    "    for url in urls:\n",
    "        link_urls = []\n",
    "        browser.get(url)  # load url\n",
    "        rsid = browser.find_element_by_css_selector('.firstHeading').text\n",
    "        links_elements = browser.find_elements_by_partial_link_text('PMID')\n",
    "\n",
    "        # get the URLs to the PubMed pages\n",
    "        for link in links_elements:\n",
    "            tempUrl = link.get_attribute('href')\n",
    "            link_urls.append(tempUrl.split(\"?\")[0])\n",
    "        print(len(link_urls))\n",
    "        # follow each link element to PubMed\n",
    "        for element in link_urls:\n",
    "            browser.get(element)\n",
    "            time.sleep(2)\n",
    "            citation_title = browser.find_element(By.XPATH, \"//meta[@name='citation_title']\")\n",
    "            article_title = citation_title.get_attribute('content')\n",
    "            citation_authors = browser.find_element(By.XPATH, \"//meta[@name='citation_authors']\")\n",
    "            article_authors = citation_authors.get_attribute('content')\n",
    "            citation_journal_title = browser.find_element(By.XPATH, \"//meta[@name='citation_journal_title']\")\n",
    "            article_citation = citation_journal_title.get_attribute('content')\n",
    "            # abs = browser.find_element_by_class_name('abstract')\n",
    "\n",
    "            rsid_list.append(rsid)\n",
    "            all_article_title.append(article_title)\n",
    "            all_article_citation.append(article_citation)\n",
    "            all_article_authors.append(article_authors)\n",
    "            # all_article_abstract.append(abs)\n",
    "            all_article_links.append(element)\n",
    "\n",
    "        # store the information\n",
    "    df = pd.DataFrame()\n",
    "    df['rsid'] = rsid_list\n",
    "    df['article_title'] = all_article_title\n",
    "    df['article_citation'] = all_article_citation\n",
    "    df['article_authors'] = all_article_authors\n",
    "    # df['article_abstract'] = all_article_abstract\n",
    "    df['link'] = all_article_links\n",
    "\n",
    "    df = df.drop_duplicates()\n",
    "\n",
    "    df.index = range(len(df.index))\n",
    "    return df\n",
    "writer = pd.ExcelWriter('./results/interesting_genes.xlsx', engine='xlsxwriter')\n",
    "\n",
    "for key in urlDictionary:\n",
    "    abstracts_df = scrape_abstracts(urlDictionary[key])\n",
    "    abstracts_df.to_excel(writer, sheet_name=key)\n",
    "writer.save()\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
